sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java

```diff
@@ -569,10 +613,353 @@ public boolean getValidate() {
       public Boolean getFlattenResults() {
         return flattenResults;
       }
+
+      private BigQueryServices getBigQueryServices() {
+        if (testBigQueryServices != null) {
+          return testBigQueryServices;
+        } else {
+          return new BigQueryServicesImpl();
+        }
+      }
     }
 
     /** Disallow construction of utility class. */
     private Read() {}
+
+    static class BigQuerySource extends BoundedSource<TableRow> {
+      // The maximum number of attempts to verify temp files.
+      private static final int MAX_FILES_VERIFY_ATTEMPTS = 10;
+
+      // The maximum number of retries to poll the status of an BigQuery job.
+      private static final int MAX_JOB_STATUS_POLL_RETRIES = 20;
+
+      // The initial backoff for verifying temp files.
+      private static final long INITIAL_FILES_VERIFY_BACKOFF_MILLIS = TimeUnit.SECONDS.toMillis(1);
+
+      private final String jobIdToken;
+      @Nullable private final String jsonTable;
+      @Nullable private final String query;
+      @Nullable private final String jsonQueryTempTable;
+      @Nullable private Boolean flattenResults;
+      private final BigQueryServices bqServices;
+
+      static BigQuerySource fromQuery(
+          String jobIdToken,
+          String query,
+          String jsonQueryTempTable,
+          Boolean flattenResults,
+          BigQueryServices bqServices) {
+        return new BigQuerySource(
+            checkNotNull(jobIdToken, "jobIdToken"),
+            null, // jsonTable
+            checkNotNull(query, "query"),
+            checkNotNull(jsonQueryTempTable, "jsonQueryTempTable"),
+            checkNotNull(flattenResults, "flattenResults"),
+            checkNotNull(bqServices, "bqServices"));
+      }
+
+      static BigQuerySource fromTable(
+          String jobIdToken,
+          String jsonTable,
+          BigQueryServices bqServices) {
+        return new BigQuerySource(
+            checkNotNull(jobIdToken, "jobIdToken"),
+            checkNotNull(jsonTable, "jsonTable"),
+            null, // query
+            null, // jsonQueryTempTable
+            null, // flattenResults
+            checkNotNull(bqServices, "bqServices"));
+      }
```

Comment:
Consider whether it's worth separating into `BigQueryTableSource` and `BigQueryQuerySource` (name?) to make this code less ambiguous. (They could both use the same reader of course.)